Add geometric figures and utility classes

MaritaPasha opened the pull request:
- Created interfaces AreaCalculator and Draw for calculating area and drawing figures
- Added Color enum with predefined color constants
- Implemented abstract class Figure with color property
- Developed concrete figure classes: Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid
- Added ColorSupplier to provide random colors
- Added FigureSupplier to generate random and default figures
- Implemented Main class demonstrating creation and drawing of figures

MaritaPasha committed:
Add geometric figures and utility classes

- Created interfaces AreaCalculator and Draw for calculating area and drawing figures
- Added Color enum with predefined color constants
- Implemented abstract class Figure with color property
- Developed concrete figure classes: Square, Rectangle, RightTriangle, Circle, IsoscelesTrapezoid
- Added ColorSupplier to provide random colors
- Added FigureSupplier to generate random and default figures
- Implemented Main class demonstrating creation and drawing of figures

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ColorSupplier.java:
```diff
@@ -0,0 +1,13 @@
+    public String getRandomColor() {
+        int index = random.nextInt(COLORS.length);
+        return COLORS[index].name().toLowerCase();
+    }
```
Issue: The getRandomColor() method returns a String (the color name in lowercase), but according to the checklist, it should return a Color enum value. Please update the method to return a Color instead of a String.

Checklist reference: 'ColorSupplier#getRandomColor() should return a Color enum, not a String.'

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+    private final ColorSupplier colorSupplier = new ColorSupplier();
+
+    public Figure getRandomFigure() {
+        String color = colorSupplier.getRandomColor();
```
Issue: colorSupplier.getRandomColor() returns a String, but according to the checklist, it should return a Color enum. Update the method and usages to work with the Color enum.

Checklist reference: 'ColorSupplier#getRandomColor() should return a Color enum, not a String.'

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+        int figureType = random.nextInt(FIGURE_TYPES_COUNT);
+
+        return switch (figureType) {
+            case 0 -> generateSquare(color);
+            case 1 -> generateRectangle(color);
+            case 2 -> generateRightTriangle(color);
+            case 3 -> generateCircle(color);
+            case 4 -> generateTrapezoid(color);
```
Issue: All generate* methods use String for color, but the checklist requires using the Color enum throughout the code. Update all relevant methods and constructors to use Color instead of String for color.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public Figure getDefaultFigure() {
+        final String color = Color.WHITE.name().toLowerCase();
```
Issue: The default color is set as a String using Color.WHITE.name().toLowerCase(), but the checklist requires using the Color enum. Pass Color.WHITE directly and update constructors accordingly.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+        double lowerBase = getRandomValue() + MIN_BASE_DIFFERENCE;
+        double upperBase = random.nextDouble() * (lowerBase - MIN_BASE_DIFFERENCE) + MIN_VALUE;
+        double height = getRandomValue();
+        return new IsoscelesTrapezoid(color, lowerBase, upperBase, height);
```
Issue: All generate* methods accept and pass color as a String. Update these methods and the corresponding figure constructors to use the Color enum instead of String.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/IsoscelesTrapezoid.java:
```diff
@@ -0,0 +1,33 @@
+
+    public IsoscelesTrapezoid(String color, double lowerBaseLength,
+                              double upperBaseLength, double heightLength) {
+        super(color);
```
Issue: The constructor uses String for the color parameter and passes it to the superclass. According to the checklist, you should use the Color enum throughout the code, not String for color. Update the parameter type and the call to super accordingly.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor replied:
Related to the above: Ensure that the superclass `Figure` also expects a `Color` enum, not a `String`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -0,0 +1,29 @@
+    private final double secondSide;
+
+    public Rectangle(String color, double firstSide, double secondSide) {
+        super(color);
```
Issue: The constructor uses String for the color parameter and passes it to the superclass. According to the checklist, you should use the Color enum throughout the code, not String for color. Update the parameter type and the call to super accordingly.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor replied:
Related to the above: Ensure that the superclass `Figure` also expects a `Color` enum, not a `String`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/RightTriangle.java:
```diff
@@ -0,0 +1,29 @@
+    private final double secondLeg;
+
+    public RightTriangle(String color, double firstLeg, double secondLeg) {
+        super(color);
```
Issue: The constructor uses String for the color parameter and passes it to the superclass. According to the checklist, you should use the Color enum throughout the code, not String for color. Update the parameter type and the call to super accordingly.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor replied:
Related to the above: Ensure that the superclass `Figure` also expects a `Color` enum, not a `String`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Square.java:
```diff
@@ -0,0 +1,26 @@
+    private final double side;
+
+    public Square(String color, double side) {
+        super(color);
```
Issue: The constructor uses String for the color parameter and passes it to the superclass. According to the checklist, you should use the Color enum throughout the code, not String for color. Update the parameter type and the call to super accordingly.

Checklist reference: 'Use the Color enum throughout the code, not String for color.'

mateacademy-ai-mentor replied:
Related to the above: Ensure that the superclass `Figure` also expects a `Color` enum, not a `String`.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Thank you for your hard work! Before your code can be approved, there are a few critical issues that need to be addressed. The main problem is the use of `String` for color throughout your code, instead of the required `Color` enum. This affects constructors, suppliers, and method signatures in multiple files. Additionally, your `ColorSupplier#getRandomColor()` method returns a `String` instead of a `Color` enum, which does not meet the core requirements. Please update all relevant code to use the `Color` enum for color properties and parameters, and ensure that all suppliers and constructors are consistent with this change. Once these issues are fixed, your solution will be much closer to meeting all the requirements—keep going, you're almost there! 🚀

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

MaritaPasha committed:
Refactor FigureSupplier to use Color enum instead of String for color

- Updated all methods to accept and pass Color enum values
- Fixed getRandomFigure() and getDefaultFigure() to work with Color
- Ensured consistency with Color enum throughout the class
- Prepared constructors to accept Color instead of String

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Circle.java:
```diff
@@ -0,0 +1,25 @@
+public class Circle extends Figure {
+    private final double radius;
+
+    public Circle(String color, double radius) {
```
Issue: The constructor uses `String color`, but the requirements specify that the `Color` enum should be used for color properties and parameters. Please update the parameter type to `Color`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Circle.java:
```diff
@@ -0,0 +1,25 @@
+    private final double radius;
+
+    public Circle(String color, double radius) {
+        super(color);
```
Related to the above: Ensure that the superclass `Figure` also expects a `Color` enum, not a `String`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Circle.java:
```diff
@@ -0,0 +1,25 @@
+                + getArea()
+                + " sq. units, radius: "
+                + radius
+                + " units, color: "
+                + getColor());
```
Issue: The `draw` method uses `getColor()`, which should return a `Color` enum, not a `String`. Please ensure that the color property and its getter are updated to use the `Color` enum throughout your codebase.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Figure.java:
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax;
+
+public abstract class Figure implements AreaCalculator, PrintableFigure {
+    protected String color;
```
Issue: The `color` property is declared as `String`, but the requirements specify that it should be of type `Color` (the enum). Please update this to `protected Color color;`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Figure.java:
```diff
@@ -0,0 +1,13 @@
+public abstract class Figure implements AreaCalculator, PrintableFigure {
+    protected String color;
+
+    public Figure(String color) {
```
Issue: The constructor parameter should be of type `Color`, not `String`. Please update the parameter type.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Figure.java:
```diff
@@ -0,0 +1,13 @@
+    protected String color;
+
+    public Figure(String color) {
+        this.color = color;
```
Related to the above: Assign the `Color` enum value to the `color` property.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Figure.java:
```diff
@@ -0,0 +1,13 @@
+        this.color = color;
+    }
+
+    public String getColor() {
```
Issue: The getter should return `Color`, not `String`. Please update the return type and method signature.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+    public Figure getDefaultFigure() {
+        final Color color = Color.WHITE;
+        final int radius = 10;
+        return new Circle(color.name(), radius);
```
Issue: The `Circle` constructor is called with `color.name()`, which is a `String`. The requirements specify that the `Color` enum should be used directly. Please pass `color` instead of `color.name()`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+
+    private Square generateSquare(Color color) {
+        double side = getRandomValue();
+        return new Square(color.name(), side);
```
Issue: The `Square` constructor is called with `color.name()`, which is a `String`. Please pass the `Color` enum value directly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+    private Rectangle generateRectangle(Color color) {
+        double width = getRandomValue();
+        double height = getRandomValue();
+        return new Rectangle(color.name(), width, height);
```
Issue: The `Rectangle` constructor is called with `color.name()`, which is a `String`. Please pass the `Color` enum value directly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+    private RightTriangle generateRightTriangle(Color color) {
+        double firstLeg = getRandomValue();
+        double secondLeg = getRandomValue();
+        return new RightTriangle(color.name(), firstLeg, secondLeg);
```
Issue: The `RightTriangle` constructor is called with `color.name()`, which is a `String`. Please pass the `Color` enum value directly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+
+    private Circle generateCircle(Color color) {
+        double radius = getRandomValue();
+        return new Circle(color.name(), radius);
```
Issue: The `Circle` constructor is called with `color.name()`, which is a `String`. Please pass the `Color` enum value directly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/FigureSupplier.java:
```diff
@@ -0,0 +1,66 @@
+        double lowerBase = getRandomValue() + MIN_BASE_DIFFERENCE;
+        double upperBase = random.nextDouble() * (lowerBase - MIN_BASE_DIFFERENCE) + MIN_VALUE;
+        double height = getRandomValue();
+        return new IsoscelesTrapezoid(color.name(), lowerBase, upperBase, height);
```
Issue: The `IsoscelesTrapezoid` constructor is called with `color.name()`, which is a `String`. Please pass the `Color` enum value directly.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/IsoscelesTrapezoid.java:
```diff
@@ -0,0 +1,33 @@
+    private final double upperBaseLength; // довжина верхньої основи має бути менша
+    private final double heightLength;
+
+    public IsoscelesTrapezoid(String color, double lowerBaseLength,
```
Issue: The constructor uses `String color`, but the requirements specify that the `Color` enum should be used for color properties and parameters. Please update the parameter type to `Color`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/IsoscelesTrapezoid.java:
```diff
@@ -0,0 +1,33 @@
+                + upperBaseLength
+                + " units, heightLength: "
+                + heightLength + " units, color: "
+                + getColor());
```
Issue: The `draw` method uses `getColor()`, which should return a `Color` enum, not a `String`. Please ensure that the color property and its getter are updated to use the `Color` enum throughout your codebase.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -0,0 +1,29 @@
+    private final double firstSide;
+    private final double secondSide;
+
+    public Rectangle(String color, double firstSide, double secondSide) {
```
Issue: The constructor uses `String color`, but the requirements specify that the `Color` enum should be used for color properties and parameters. Please update the parameter type to `Color`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -0,0 +1,29 @@
+                + firstSide
+                + " units, width: "
+                + secondSide + " units, color: "
+                + getColor());
```
Issue: The `draw` method uses `getColor()`, which should return a `Color` enum, not a `String`. Please ensure that the color property and its getter are updated to use the `Color` enum throughout your codebase.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/RightTriangle.java:
```diff
@@ -0,0 +1,29 @@
+    private final double firstLeg;
+    private final double secondLeg;
+
+    public RightTriangle(String color, double firstLeg, double secondLeg) {
```
Issue: The constructor uses `String color`, but the requirements specify that the `Color` enum should be used for color properties and parameters. Please update the parameter type to `Color`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/RightTriangle.java:
```diff
@@ -0,0 +1,29 @@
+                + firstLeg
+                + " units, secondLeg: "
+                + secondLeg + " units, color: "
+                + getColor());
```
Issue: The `draw` method uses `getColor()`, which should return a `Color` enum, not a `String`. Please ensure that the color property and its getter are updated to use the `Color` enum throughout your codebase.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Square.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final double side;
+
+    public Square(String color, double side) {
```
Issue: The constructor uses `String color`, but the requirements specify that the `Color` enum should be used for color properties and parameters. Please update the parameter type to `Color`.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/Square.java:
```diff
@@ -0,0 +1,26 @@
+                + " sq. units, side: "
+                + side
+                + " units, color: "
+                + getColor());
```
Issue: The `draw` method uses `getColor()`, which should return a `Color` enum, not a `String`. Please ensure that the color property and its getter are updated to use the `Color` enum throughout your codebase.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Thank you for your hard work! 😊 However, there are several critical issues that must be addressed before your Pull Request can be approved. The most important problem is that all figure classes (and the superclass `Figure`) use `String` for the color property, constructor, and getter, while the requirements clearly state that the `Color` enum should be used throughout. Additionally, constructors are being called with `color.name()` (a `String`) instead of passing the `Color` enum directly, and the `ColorSupplier#getRandomColor()` method should return a `String` (the color name), not a `Color` enum. The use of an abstract `Figure` class is also not allowed by the requirements. Please update your code to use the `Color` enum for all color properties, parameters, and methods, refactor the constructors and usages accordingly, and remove the abstract `Figure` class. Once these critical issues are fixed, your solution will meet the core requirements. Keep going—you’re almost there! 🚀

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

MaritaPasha committed:
Refactor FigureSupplier to use Color enum instead of String for color

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/logic/Figure.java:
```diff
@@ -0,0 +1,15 @@
+
+import core.basesyntax.Color;
+
+public abstract class Figure implements AreaCalculator, PrintableFigure {
```
Issue: The `Figure` class is declared as `abstract`, which is not allowed according to the task requirements and checklist. Please remove the `abstract` keyword so that `Figure` is a concrete class.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Thank you for your submission! 🚀 Your solution is well-structured and meets almost all the requirements for the task. All figure classes, suppliers, interfaces, and the main logic are implemented correctly, and your code demonstrates a solid understanding of OOP principles and clean coding practices.

However, there is one critical issue that you must address in future submissions:  
**The `Figure` class is declared as `abstract`, which is explicitly forbidden by the task requirements.** The checklist and task description state that you should NOT use abstract classes to set behavior for classes. Please remove the `abstract` modifier from the `Figure` class to fully comply with the requirements.

All other aspects of your code are correct, and your implementation is functional and clear. I am approving your pull request to help you move forward with your learning, but please make sure to review the comments and fix the abstract class issue in your next iteration. Addressing such critical requirements is important for both passing the task and improving your coding discipline.

Keep up the good work, and don't hesitate to revisit the requirements or ask questions if anything is unclear. Every mistake is an opportunity to learn and grow as a developer! 👍

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
